### DISCUSSION TITLE
How to cache an aggregate root?

### DISCUSSION THREAD

**Discussion by yqszt** [abpframework/abp#12378] [Q&A] [answered]
i have a question about how to cache an aggregate root. 
generally we create a cache item for our entity, but if we want to cache an aggregate root, the root may contains many other entity or entity list. so need i to create cache item class for each entity? if yes, we will have many object map work. the automapper module should be import to domain project?
although aggregate root is serializable, but the entity fields have special busniess logic, so these fields maybe protected, we can cahce an aggregate root to redis, but when we get this cache, the aggregate root cannot be correct deserialize. 
what is the best practice to cache aggregate root?

**Comment by EngincanV** [comment, marked as answer]
> need i to create cache item class for each entity? if yes, we will have many object map work. the automapper module should be import to domain project?

- I think it's better to define cache item classes. (Example: [BookCacheItem](https://docs.abp.io/en/abp/latest/Caching#idistributedcache-tcacheitem-interface) class from Caching documentation) Because by doing this you can cache certain values and get the other values from the database. (And I think it's also good for the SOC perspective)
- You don't need to import the `AutoMapperModule` to your domain module. Your domain module should not depend on it. You can make the related mapping configurations from the application layer.

> what is the best practice to cache aggregate root?
- Defining a cache item class and only include the properties that you want to cache (most frequently asked/queried or all of them if required).
- You can read the [Caching](https://docs.abp.io/en/abp/latest/Caching) documentation of ABP, it already shows a sample about creating a cache item class and using it when caching.

**Reply by yqszt** [reply]
thank you so much :)

**Reply by EngincanV** [reply]
You're welcome :)
